lib/backend/k8s/resources/resources.go

```diff
@@ -100,3 +102,74 @@ func SetCalicoMetadataFromK8sAnnotations(calicoRes Resource, k8sRes Resource) {
 		}
 	}
 }
+
+// Store Calico Metadata in the in the k8s resource annotations for CRD backed resources.
+// This should store all Metadata except for those stored in Annotations and Labels and
+// store them in annotations.
+func StoreCalicoMetadataInK8sAnnotations(res Resource) error {
+	metadataBytes, err := json.Marshal(res.GetObjectMeta().(*metav1.ObjectMeta))
+	if err != nil {
+		return err
+	}
+	annotations := res.GetObjectMeta().GetAnnotations()
+	if len(annotations) == 0 {
+		annotations = make(map[string]string)
+	}
+	annotations[metadataAnnotation] = string(metadataBytes)
+	res.GetObjectMeta().SetAnnotations(annotations)
+
+	// Make sure to clear out all of the Calico Metadata out of the ObjectMeta except for
+	// Name, Namespace, and ResourceVersion.
+	res.GetObjectMeta().SetSelfLink("")
+	res.GetObjectMeta().SetUID("")
+	res.GetObjectMeta().SetGeneration(int64(0))
+	var emptyTime metav1.Time
+	res.GetObjectMeta().SetCreationTimestamp(emptyTime)
+	res.GetObjectMeta().SetDeletionTimestamp(nil)
+	res.GetObjectMeta().SetDeletionGracePeriodSeconds(nil)
+	res.GetObjectMeta().SetInitializers(nil)
+	var emptyStringSlice []string
+	res.GetObjectMeta().SetFinalizers(emptyStringSlice)
+	var emptyReferences []metav1.OwnerReference
+	res.GetObjectMeta().SetOwnerReferences(emptyReferences)
+	res.GetObjectMeta().SetClusterName("")
```

Comment:
All of this field copying fields like a maintenance hassle - especially if k8s decide to add more fields to the Metadata.

I think a better a more maintainable way to do this would be to use the DeepCopyInto() method defined on the ObjectMeta struct.  You could create your own local "empty" ObjectMeta with just the fields that should be included:  ie. the name, namespace, labels and annotations - and then use DeepCopyInto to copy over the Metadata in the resource.  This is more resilient to new fields being added.

Reply:
Same goes for both of these conversion functions.

Reply:
I think you should also avoid storing duplicate copies of everything - so I think the ObjectMeta that is stored in the annotation should have the Name, Namespace, TypeMeta, Annotations and Labels removed before marshaling into json.

Obviously you'll need to add those fields back in the other function.

Reply:
I thought about using the `DeepCopyInto` method to create the metadata, but the interface definitions of the `Resource` do not have any methods to replace the `ObjectMeta` struct do they? The issue is I think we need to use the resources for these since the concrete value of each resource is different depending on the type.

Reply:
No you can't replace the Metadata unfortunately - but you can deep copy into it, and the GetObjectMetadata() returns you the pointer to the ObjectMeta.

Reply:
Good point, I'll fix that.